File: 2024/10.py
import time

from utils import readInput, DIRECTIONS

dirs = {i:d for i,d in enumerate(DIRECTIONS)}

# TRY DFS to speed up

# def dfs(graph, start_node, visited=None):
#   if visited is None:
#     visited = set()

#   visited.add(start_node)
#   print(start_node)

#   for neighbor in graph[start_node]:
#     if neighbor not in visited:
#       dfs(graph, neighbor, visited)

def loadInput(filename: str) -> tuple[dict[complex, int], list[complex]]:
    lines = readInput(filename)
    starts: list[complex] = []
    map: dict[complex, int] = {}
    for y in range(len(lines)):
        for x in range(len(lines[0])):
            if lines[y][x] == ".":
                map[complex(x, y)] = 0
            elif lines[y][x] == "0":
                starts.append(complex(x, y))
                map[complex(x, y)] = int(lines[y][x])
            else:
                map[complex(x, y)] = int(lines[y][x])
    return map, starts

def part1(map: dict[complex, int], starts: list[complex]):
    score = 0
    for start in starts:
        queue = [(start, 0)]
        visited = [start]
        pos = start
        while len(queue) > 0:
            pos = queue.pop()
            visited.append(pos[0])
            if pos[1] == 9:
                score += 1
            for i in range(4):
                npos = pos[0] + dirs[i]
                if npos in map and npos not in visited:
                    if (map[npos] - pos[1]) == 1:
                        queue.append((npos, map[npos]))
    print (f"🎄 Part 1: {score}", end="")


def part2(map: dict[complex, int], starts: list[complex]):
    rate: dict[complex, int] = {}
    for start in starts:
        queue = [(start, 0)]
        pos = start
        while len(queue) > 0:
            pos = queue.pop()
            if pos[1] == 9:
                rate[pos[0]] = rate.get(pos[0], 0) + 1
            for i in range(4):
                npos = pos[0] + dirs[i]
                if npos in map:
                    if (map[npos] - pos[1]) == 1:
                        queue.append((npos, map[npos]))
    print (f"🎄🎅 Part 2: {sum(rate.values())}", end="")

    
if __name__ == '__main__':
    title = "Day 10: Hoof It"
    sub = "❄ "*(len(title)//2+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_10.txt")

    t0 = time.time()
    part1(*inputs)
    print (f" - {time.time()-t0:.5f}")
    
    t0 = time.time()
    part2(*inputs)
    print (f" - {time.time()-t0:.5f}")
    

File: 2024/10.jl
dirs = Dict(0=>Complex(0, -1), 1=>Complex(1, 0), 2=>Complex(0, 1), 3=>Complex(-1, 0))

function loadInput(filename::String)
    lines = readlines(filename)
    starts = []
    map = Dict{Complex, Int}()
    for y in eachindex(lines)
        for x in eachindex(lines[y])
            if lines[y][x] == '.'
                map[Complex(x, y)] = 0
            elseif lines[y][x] == '0'
                push!(starts,  Complex(x, y))
                map[Complex(x, y)] = parse(Int, lines[y][x])
            else
                map[Complex(x, y)] = parse(Int, lines[y][x])
            end
        end
    end
    return map, starts
end
    
function part1(map, starts)
    score = 0
    for start in starts
        queue = [(start, 0)]
        visited = [start]
        pos = start
        while length(queue) > 0
            pos = pop!(queue)
            push!(visited, pos[1])
            if pos[2] == 9
                score += 1
            end

            for i in 0:3
                npos = pos[1] + dirs[i]
                if npos in keys(map) && !(npos in visited)
                    if (map[npos] - pos[2]) == 1
                        push!(queue, (npos, map[npos]))
                    end
                end
            end
        end
    end

    println("🎄 Part 1:  $(score)")
end

function part2(map, starts)
    rate = Dict{Complex, Int}()
    for start in starts
        queue = [(start, 0)]
        pos = start
        while length(queue) > 0
            pos = pop!(queue)
            if pos[2] == 9
                rate[pos[1]] = get(rate, pos[1], 0) + 1
            end
            for i in 0:3
                npos = pos[1] + dirs[i]
                if npos in keys(map)
                    if (map[npos] - pos[2]) == 1
                        push!(queue, (npos, map[npos]))
                    end
                end
            end
        end
    end

    println("🎄🎅 Part 2: $(sum(values(rate)))")
end

function main()
    title = "Day 10: Hoof It"
    sub = "❄ "^(length(title)÷2 + 2)
    
    println()
    println(" $title ")
    println(sub)
    
    inputs = loadInput("input_10.txt")
    @time part1(inputs...)
    @time part2(inputs...)
end

main()